kubernetes/resource_kubernetes_validating_webhook_configuration.go

```diff
@@ -0,0 +1,310 @@
+							Optional:    true,
+							MaxItems:    1,
+							Elem: &schema.Resource{
+								Schema: labelSelectorFields(true),
+							},
+						},
+						"rule": {
+							Type:        schema.TypeList,
+							Description: "A rule describes what operations on what resources/subresources the webhook cares about. The webhook cares about an operation if it matches _any_ Rule. However, in order to prevent ValidatingAdmissionWebhooks and MutatingAdmissionWebhooks from putting the cluster in a state which cannot be recovered from without completely disabling the plugin, ValidatingAdmissionWebhooks and MutatingAdmissionWebhooks are never called on admission requests for ValidatingWebhookConfiguration and MutatingWebhookConfiguration objects.",
+							Required:    true,
+							MinItems:    1,
+							Elem: &schema.Resource{
+								Schema: ruleWithOperationsFields(),
+							},
+						},
+						"side_effects": {
+							Type:        schema.TypeString,
+							Description: `states whether this webhook has side effects. Acceptable values are: None, NoneOnDryRun (webhooks created via v1beta1 may also specify Some or Unknown). Webhooks with side effects MUST implement a reconciliation system, since a request may be rejected by a future step in the admission change and the side effects therefore need to be undone. Requests with the dryRun attribute will be auto-rejected if they match a webhook with sideEffects == Unknown or Some.`,
+							Optional:    true,
+						},
+						"timeout_seconds": {
+							Type:        schema.TypeInt,
+							Description: "TimeoutSeconds specifies the timeout for this webhook. After the timeout passes, the webhook call will be ignored or the API call will fail based on the failure policy. The timeout value must be between 1 and 30 seconds. Default to 10 seconds.",
+							Default:     10,
+							Optional:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceKubernetesValidatingWebhookConfigurationCreate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	conn, err := meta.(KubeClientsets).MainClientset()
+	if err != nil {
+		return err
+	}
+
+	cfg := admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: expandMetadata(d.Get("metadata").([]interface{})),
+		Webhooks:   expandValidatingWebhooks(d.Get("webhook").([]interface{})),
+	}
+
+	log.Printf("[INFO] Creating new ValidatingWebhookConfiguration: %#v", cfg)
+
+	res := &admissionregistrationv1.ValidatingWebhookConfiguration{}
+
+	if useAdmissionregistrationV1beta1(conn) {
```

Comment:
@dak1n1 @alexsomesan Can I get your thoughts on this approach to falling back to the old version of the API. I'm trying to avoid duplicating tons of code with the types from `v1beta1` like I had to do with the HPA. Alternatives here are to use unsafe.Pointer (😱) or duplicate all the code to support both types. If there is an easier way to do this that I've missed please let me know.


This code works because `v1` and `v1beta1` have identical struct fields, it is just the struct tags that have changed. Here is the diff:

```
~/d/k/a/admissionregistration ❯❯❯ diff v1/types.go v1beta1/types.go
2c2
< Copyright 2019 The Kubernetes Authors.
---
> Copyright 2017 The Kubernetes Authors.
17c17
< package v1
---
> package v1beta1
117a118
> // Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead.
147a149
> // Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead.
195c197
< 	// allowed values are Ignore or Fail. Defaults to Fail.
---
> 	// allowed values are Ignore or Fail. Defaults to Ignore.
212c214
< 	// Defaults to "Equivalent"
---
> 	// Defaults to "Exact"
277c279
< 	// Acceptable values are: None, NoneOnDryRun (webhooks created via v1beta1 may also specify Some or Unknown).
---
> 	// Acceptable values are: Unknown, None, Some, NoneOnDryRun
281,282c283,285
< 	// sideEffects == Unknown or Some.
< 	SideEffects *SideEffectClass `json:"sideEffects" protobuf:"bytes,6,opt,name=sideEffects,casttype=SideEffectClass"`
---
> 	// sideEffects == Unknown or Some. Defaults to Unknown.
> 	// +optional
> 	SideEffects *SideEffectClass `json:"sideEffects,omitempty" protobuf:"bytes,6,opt,name=sideEffects,casttype=SideEffectClass"`
288c291
< 	// Default to 10 seconds.
---
> 	// Default to 30 seconds.
299c302,304
< 	AdmissionReviewVersions []string `json:"admissionReviewVersions" protobuf:"bytes,8,rep,name=admissionReviewVersions"`
---
> 	// Default to `['v1beta1']`.
> 	// +optional
> 	AdmissionReviewVersions []string `json:"admissionReviewVersions,omitempty" protobuf:"bytes,8,rep,name=admissionReviewVersions"`
324c329
< 	// allowed values are Ignore or Fail. Defaults to Fail.
---
> 	// allowed values are Ignore or Fail. Defaults to Ignore.
341c346
< 	// Defaults to "Equivalent"
---
> 	// Defaults to "Exact"
406c411
< 	// Acceptable values are: None, NoneOnDryRun (webhooks created via v1beta1 may also specify Some or Unknown).
---
> 	// Acceptable values are: Unknown, None, Some, NoneOnDryRun
410,411c415,417
< 	// sideEffects == Unknown or Some.
< 	SideEffects *SideEffectClass `json:"sideEffects" protobuf:"bytes,6,opt,name=sideEffects,casttype=SideEffectClass"`
---
> 	// sideEffects == Unknown or Some. Defaults to Unknown.
> 	// +optional
> 	SideEffects *SideEffectClass `json:"sideEffects,omitempty" protobuf:"bytes,6,opt,name=sideEffects,casttype=SideEffectClass"`
417c423
< 	// Default to 10 seconds.
---
> 	// Default to 30 seconds.
428c434,436
< 	AdmissionReviewVersions []string `json:"admissionReviewVersions" protobuf:"bytes,8,rep,name=admissionReviewVersions"`
---
> 	// Default to `['v1beta1']`.
> 	// +optional
> 	AdmissionReviewVersions []string `json:"admissionReviewVersions,omitempty" protobuf:"bytes,8,rep,name=admissionReviewVersions"`
```

Reply:
How about making a helper function that will return the WebhookConfig for the appropriate API version? Something like this:

```
// createWebhookConfiguration will determine the kubernetes API version and generate a compatible webhook configuration.
func createWebhookConfiguration(d *schema.ResourceData, meta interface{}) (interface{}, error) {

	conn, err := meta.(KubeClientsets).MainClientset()
	if err != nil {
		return nil, err
	}

	if useAdmissionregistrationV1beta1(conn) {
		return &admissionregistrationv1beta1.ValidatingWebhookConfiguration{}, nil
	} else {
		return &admissionregistrationv1.ValidatingWebhookConfiguration{}, nil
	}
}
```

Then you can keep the CRUD functions agnostic of the API version. For example:

```
func resourceKubernetesValidatingWebhookConfigurationRead(d *schema.ResourceData, meta interface{}) error {

	conn, err := meta.(KubeClientsets).MainClientset()
	if err != nil {
		return err
	}

	name := d.Id()

	cfg, err := createWebhookConfiguration(d, meta)
	if err != nil {
		return err
	}
```

Then you'd use the variable `cfg` without needing to know its API version. If that isn't possible (for example, if you need to run an API version-specific operation on it), then we could have the helper function return an object that is version-specific to run the CRUD operations on. Just an idea.

Reply:
I tried something like this initially, but I ran into the problem that all the expander and flattener functions are using `admissionregistration/v1` in their signatures, and you can't cast or convert types from one API version to the other, even if they are `interface{}`. The problem I'm trying to avoid is duplicating all those functions so we have signatures for both v1 and v1beta1. 😢

Reply:
Today I was trying to figure out if we could solve this duplication issue by creating a kind of compatibility layer between the two API versions, similar to what I did in a different project.

My compatibility layer in that other project was basically a struct that contained an interface. And that interface could hold either of the two objects I was working with. That allowed my functions to accept either one of those similar objects interchangeably. The two interchangeable objects were an AWS S3 client and an AWS S3 mock client, but the same principle could apply here if we had a struct that holds an interface resembling the `ValidatingWebhookConfiguration`. That way you could squeeze either a `v1.ValidatingWebhookConfiguration` or a `v1beta.ValidatingWebhookConfiguration` into the same struct, and use that struct in version-agnostic functions.

Here's where I made that compatibility thing, in case it helps to see what I'm talking about. https://github.com/dak1n1/managed-velero-operator/blob/89f07878d7eebe4cfea721246a3ef635f8fca9f2/pkg/s3/client.go#L30-L55

I tried coding up something like that today, but it got pretty messy and I think I got in over my head with that. It looks like more trouble than it's worth, but I wanted to share it in case it sparked any ideas. Some code duplication might be more straightforward to maintain, compared to creating a compatibility layer abstraction. :thinking: 

It might even be neater to have different CRUD functions for the v1 API vs the v1beta API. That way it'll be really quick and easy to remove when the backwards-compatibility is no longer needed. I think as long as the code duplication is separated out into easy-to-understand (and easy-to-remove) blocks, it could be good to just duplicate it.